src/raft.rs

```diff
@@ -79,6 +80,77 @@ pub struct SoftState {
     pub raft_state: StateRole,
 }
 
+/// UncommittedState is used to keep track of imformation of uncommitted
+/// log entries on 'leader' node
+struct UncommittedState {
+    /// Specify maximum of uncommited entry size.
+    /// When this limit is reached, all proposals to append new log will be dropped
+    max_uncommitted_size: usize,
+
+    /// Record current uncommitted entries size.
+    uncommitted_size: usize,
+
+    /// Record last committed log index when node becomes leader from candidate.
+    /// See https://github.com/tikv/raft-rs/pull/398#discussion_r502417531 for more detail
+    last_committed_index: u64,
+}
+
+impl UncommittedState {
+    #[inline]
+    pub fn set_uncommitted_size(&mut self, size: usize) {
+        self.uncommitted_size = size;
+    }
+
+    #[inline]
+    pub fn set_last_committed_index(&mut self, idx: u64) {
+        self.last_committed_index = idx;
+    }
+
+    #[inline]
+    pub fn is_no_limit(&self) -> bool {
+        self.max_uncommitted_size == NO_LIMIT as usize
+    }
+
+    pub fn maybe_increase_uncommitted_size(&mut self, ents: &[Entry]) -> bool {
+        // fast path
+        if self.is_no_limit() {
+            return true;
+        }
+
+        let size = ents.iter().fold(0, |acc, ent| acc + ent.get_data().len());
+
+        if size + self.uncommitted_size > self.max_uncommitted_size {
+            false
+        } else {
+            self.uncommitted_size += size;
+            true
+        }
+    }
+
+    pub fn maybe_reduce_uncommitted_size(&mut self, ents: &[Entry]) -> bool {
+        // fast path
+        if self.is_no_limit() || ents.is_empty() {
+            return true;
+        }
+
+        // user may advance a 'Ready' which is generated before this node becomes leader
+        if let Some(ent) = ents.last() {
+            if ent.index <= self.last_committed_index {
+                return false;
+            }
+        }
+
+        let size = ents.iter().fold(0, |acc, ent| acc + ent.get_data().len());
```

Comment:
```suggestion
        let size = ents.iter().skip_while(|ent| ent.index <= self.last_committed_index).fold(0, |acc, ent| acc + ent.get_data().len());
```

We should compute the `size` of entries with `ent.index > self.last_committed_index`